[Q&A 미션 0단계] 피드백 부탁드립니다.

mywnajsldkf opened the pull request:
- 불필요한 파일 이동 흔적이 남아있습니다. 미리 양해 부탁드립니다. 😅
- 작업하면서 제대로 이해하고 하는 것인지 의심 드는 부분이 여럿 있었습니다. 중요한 개념들에 대해 공부해보라고 피드백 주시면 감사하겠습니다!🙇‍♀️

# 배운점
**@RequestParam VS @PathVariable**
데이터 전달위 위해 사용되는 어노테이션으로, uri를 통해 값을 전달한다. 

> @RequestParam
```
@GetMapping("/')
public String method(@RequestParam("id") String id, @RequestParam("name") String name){

// id = "wnajsldkf"
// name = "jeongin"

return "/user/profile"
```
`https://api.com?id=wnajsldkf&name=jeongin` 이렇게 이동할 것이다. 쿼리스트링으로 부르며 Get 방식의 통신에서 주로 사용된다. 

> @PathVariable
```
@GetMapping("/{id}/{name})
public String method(@PathVariable("id") String id, @PathVariable("name") String name){

// id = "wnajsldkf"
// name = "jeongin"

return "/user/profile"
```
`https://api.com/wnajsldkf/jeongin` 이렇게 이동하며 특수문자나 비밀번호 등이 입력되면 알수없는 글로 나타난다. 직관적인게 장점이다. 

@PathVariable과 @RequestParam을 함께 쓰는 경우도 있다. 
```
@GetMapping("/{id}")
public String method(@PathVariable("id") String id, @RequestParam("name") String name) {

}
```

# 생각
* 초반에 프로젝트의 방향을 제대로 이해하지 못해 어려웠습니다. 특히 SSR에 대한 개념을 생각하지 못해 잘못된 방향으로 작업했습니다. 
* SSR 경우에는 view의 역할이 전적으로 html 파일이 작업하는 것인가요?

# 더 공부가 필요한 부분
* RestController vs Controller
* Server Side Rendering Client Side Rendering
* Redirection and Forwarding
* 웹을 구성하는 전반적인 개념
* view에 대한 부분

mywnajsldkf committed:
Chore: build.gradle 업데이트

mywnajsldkf committed:
Chore: 이전 작업 내역 삭제

mywnajsldkf committed:
feat: 테이블과 매핑할 클래스 생성

mywnajsldkf committed:
feat: UserRepository 인터페이스 생성

mywnajsldkf committed:
feat: UserController 생성

- (GET `/users`) 모든 사용자 정보 가져오기
- (POST `/users`) 신규 사용자 등록하기
- (GET `/users/{id})` 사용자 정보 가져오기
- (PUT `/users/{id})` 사용자 정보 등록하기
- (DELETE `/users/{id})` 사용자 정보 삭제하기

mywnajsldkf committed:
feat: UserNotFoundException 생성

mywnajsldkf committed:
Chores: 경로 변경

mywnajsldkf committed:
Chores: 이전 작업 내역 삭제

mywnajsldkf committed:
feat: UserController 추가

- RestController -> Controller 수정
- getRegisterForm() -> 사용자 회원가입창 띄움
- registerUser() -> 사용자 회원가입 요청
- getUser() -> 사용자 목록 불러옴
- getUserProfile() -> 특정 사용자 정보를 불러옴

mywnajsldkf committed:
feat: 사용자 찾기 예외 추가

- 일치하는 사용자 아이디가 없으면 에러 메시지를 출력한다.

mywnajsldkf committed:
chores: 경로 수정 및 불필요한 코드 삭제

- 데이터베이스 개념을 적용하지 않고 ArrayList에 저장하기 때문에 데이터베이스 관련 annotation(@Entity, @Id, @GeneratedValue) 삭제
- id 반환값 수정

mywnajsldkf committed:
chores: 내장 h2 DB 경로 삭제

mywnajsldkf committed:
chores: 파일 경로 이동으로 인한 변경

mywnajsldkf committed:
chores: 경로 변경

mywnajsldkf committed:
chores: spring doc dependency 추가, 버전 변경

mywnajsldkf committed:
chores: .gitignore 변경(이동된 파일 추적 삭제)

mywnajsldkf committed:
chores: 불필요한 파일 삭제

mywnajsldkf committed:
refactor: html 코드 중복 제거

mywnajsldkf committed:
add: MvcConfig 추가

saint6839 commented on src/main/java/codesquad/User.java:
```diff
@@ -1,41 +1,59 @@
 
 public class User {
-    private Long id;
+    private Integer id;
```
id의 타입을 Long이 아닌 Integer로 설정하신 이유가 무엇인지 궁금합니다!
추가적으로 int를 사용했을때와 Integer를 사용했을 때 어떤 차이가 있을까요?

mywnajsldkf replied:
@saint6839 
등록된 회원들을 출력하는 부분에서 3번째 회원부터 id를 찾아 출력하기 위함이었습니다!
int는 기본형인 자료형이고, Integer는 기본형을 객체로 다루기 위한 wrapper 클래스입니다. 따라서 Integer로 사용할 때는 null값을 처리할 수는 있지만, 산술 연산을 하기 위해서 wrapper 클래스에서 primitive type으로 변환하는 unboxing 과정이 필요합니다. 
+) 제가 여기서는 큰 뜻을 가지고 Integer를 사용한 것은 아니고 실수로... 사용했습니다😅

saint6839 commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,55 @@
+    private static final int START_NUMBER = 3;
+    private List<User> users = new ArrayList<>();
+
+    @GetMapping("/users/join")
```
반복되는 `/users`에 대해서 컨트롤러 상단에 `@RequestMapping`을 추가하면 중복제거가 가능할 것 같습니다!

mywnajsldkf replied:
@saint6839 반영했습니다~ 감사합니다!

saint6839 commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @PostMapping("/users")
+    public String registerUser(@ModelAttribute User user) {
```
컨트롤러 메서드의 객체를 받아오는 파라미터에 별도로 어노테이션을 선언하지 않는다면 스프링부트는 디폴트로 어떤 방식을 이용해 데이터를 받아올까요?

mywnajsldkf replied:
@saint6839 
여기서는 `@ModelAttribute` 어노테이션을 말씀하실까요? 
Model 객체로 인식하고 addAttribute를 통해서 key, value 형태로 데이터를 받아올 것 같습니다.

saint6839 commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,55 @@
+    @GetMapping("/users/{userId}")
+    public String getUserProfile(@PathVariable String userId, Model model) {
+
+        User user = findUser(userId);
```
User라는 엔티티가 아닌 DTO를 생성해 사용해 보는것은 어떨까요?

mywnajsldkf replied:
@saint6839 오~~ 조언 감사합니다! 수정하여 반영했습니다.
DTO를 데이터베이스에 나오는 view와 비슷하게 이해했는데 맞을까요?

mywnajsldkf replied:
@saint6839 저번 피드백에서 이해가 짧다고 생각해 다시 재수정했습니다!

hyukjin-lee replied:
로직은 없고 데이터를 전달을 위해 담는 그릇이라고 생각하시면 됩니다~
유래가 궁금하다면 https://martinfowler.com/eaaCatalog/dataTransferObject.html 마틴파울러의 글을 추천드려요

saint6839 commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Controller
```
`@Controller`와 `@RestController`는 어떤 차이가 있을까요?

mywnajsldkf replied:
@saint6839 
`@RestController`는 `@Controller`와 `@ResponseBody`의 조합으로 Spring 4.0에서 추가되었습니다. 
`@Controller`의 역할은 Model 객체를 만들어 데이터를 담고 객체를 전달하는 View를 담당하고, `@ResponseBody`를 사용하여 HTTP Response Body에 데이터를 담아 요청을 완료합니다. 이때 HTTP Response Body에 데이터를 담는 것은 뷰를 반환하는 대신 데이터를 반환하기 때문에 RESTful 웹 서비스에 대한 응답에 매우 유용합니다. 
`@RestController`는 `@Controller`와 `@ResponseBody`를 조합하여 객체를 반환하고 객체 데이터를 해당하는 HTTP 응답에 전송합니다. 

저는 처음에 `@RestController`를 사용했다 `@Controller`로 수정했는데 그 이유는 return에서 해당하는 화면을 넘겨주었기 때문이라고 생각합니다.

saint6839 reviewed (commented):
미션하느라 수고 많으셨습니다~!!

mywnajsldkf committed:
refactor: @RequestMapping 추가하여 중복 제거

mywnajsldkf committed:
refactor: UserDTO 생성

java-saeng commented on src/main/java/codesquad/UserNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package codesquad;
+
+public class UserNotFoundException extends RuntimeException {
```
custom Exception을 사용하신 이유가 있으실까요?

mywnajsldkf replied:
@java-saeng 
저는 지금까지 예외 클래스 이름 만으로 명확하게 설명하기 위해 custom Exception을 항상 생성했습니다. 
이번에 피드백을 받아 찾아봤는데 [custom exception을 언제 써야 할까?](https://tecoble.techcourse.co.kr/post/2020-08-17-custom-exception/) 에 대한 글이 있더라구요. 

이 글을 읽고 생각한 점은 지나친 예외 생성이 오히려 가독성을 저하시킬 수 있다는 점을 알았고, 지금처럼 단순하게 메시지만을 던지는 경우에는 굳이 따로 생성하지 않고 기존에 제공하고 있는 Exception을 사용해야겠다고 생각했습니다. 오히려 이번 기회에 자바에서 제공하는 Exception에 대해 공부하도록 하겠습니다~

java-saeng commented on src/main/resources/application.properties:
```diff
@@ -4,12 +4,17 @@ handlebars.expose-session-attributes=true
 spring.h2.console.enabled=true
 #spring.datasource.url=jdbc:h2:~/java-qna;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
-spring.datasource.url=jdbc:h2:tcp://localhost/~/test
+#spring.datasource.url=jdbc:h2:tcp://localhost:9092/board:testdb;MVCC=TRUE
```
MVCC=TRUE는 어떤 의미인지 알 수 있을까요?

mywnajsldkf replied:
@java-saeng 지금은 MVCC를 사용할지 나타내는 옵션입니다. 제공하지 않아 제거하는게 맞다고 합니다~

java-saeng replied:
MVCC 가 무엇인가요??

mywnajsldkf replied:
@java-saeng 
동시 접근을 허용하는 데이터베이스에서 동시성을 제어하기 위해 사용하는 방법을 의미하는 Multi Version Concurrency Control(다중 버전 동시어 제어)의 약자로 하네요!
(사실 아직 잘 모르겠습니다! 킵 해두고 다음에 꼭 공부하도록 하겠습니다)

java-saeng replied:
넵 현재 미션에서는 중요한 개념은 아니지만 알고 쓰시는건지 궁금해서 여쭤본거였습니다~

java-saeng commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,56 @@
+
+    private UserDTO findUser(String userId) {
+        for (User user : users) {
+            if (userId.equals(user.getUserId())) {
```
TDA 법칙을 적용해보는건 어떨까요?

mywnajsldkf replied:
@java-saeng 이 부분 잘 이해가 안갑니다...! 부연설명 부탁드립니다~

hyukjin-lee replied:
Tell don't ask 법칙이라고 검색해보시면 될 것 같습니다 ~

mywnajsldkf replied:
피드백 감사합니다~! 적용해보았습니다. 
올바른지 확인해주시면 감사하겠습니다! 

https://github.com/sproutt/spring-study-qna/pull/81/files#diff-94e9fdece9afcab044426e129e8bb806c0aa6fc5df8d41b85a4f4e4adae59b58

hyukjin-lee replied:
user.isSameId(userId) 로 구현할 수 있겠죠?

java-saeng commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    private UserDTO findUser(String userId) {
+        for (User user : users) {
+            if (userId.equals(user.getUserId())) {
+                return new UserDTO(user);
```
따로 UserDTO를 반환하시는 이유가 있으실까요~?

mywnajsldkf replied:
@java-saeng 
DTO에 대한 이해가 짧았습니다.😅

재공부 후 아래 코드처럼 수정하면서 궁금한 점이 생겼는데요~
Dto를 적용하면 Entity -> Dto로 바꾸는 과정에서 중복해서 새로운 객체가 생성되는데 이 부분에 대한 고민은 따로 안해도 되는 것인가요?

```
@GetMapping("/{userId}")
    public String getUserProfile(@PathVariable String userId, Model model) {

        UserDto user = new UserDto(userService.findUser(userId));

        model.addAttribute("name", user.getName());
        model.addAttribute("email", user.getEmail());

        return "user/profile";
    }
```

java-saeng replied:
엔티티를 직접 반환할 때의 사이드 이펙트는 어떤 것이 있을까요?

hyukjin-lee replied:
객체를 생성하는 비용은 가비지 컬렉션을 어느 정도 믿고, 성능이 중요해지는 상황이 아니라면 크게 걱정하지 않으셔도 됩니다.

mywnajsldkf replied:
@java-saeng 
엔티티를 직접 반환하면 
1. 클라이언트가 Entity를 알게 되고, Entity의 변경이 클라이언트에 직접 영향을 미칩니다. 
DTO를 사용하여 클라이언트와 Entity에서 다루는 데이터를 분리해서 변경 영향 범위를 좁혀야 합니다. 
2. Entity는 해당 Entity의 모든 정보를 담고 있어, 정보가 과노출됩니다. 따라서 DTO에서는 필요한 항목만 골라서 반환해야 합니다. 
3. 다대다 방향에서 무한 반복이 일어날 수 있습니다. 서로 참조하는 객체를 계속 호출하는 경우가 발생합니다.

java-saeng commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @PostMapping()
+    public String registerUser(User user) {
+        user.setId(users.size() + START_NUMBER);
```
id 는 user 를 구분하기 위한 식별값이라고 생각하는데, users.size() + START_NUMBER로 하신 이유가 궁금해요~

mywnajsldkf replied:
@java-saeng 
id가 식별자 역할을 제대로 하지 못한 것 같습니다~ 
저는 id를 index의 역할로 사용했고, 사용자 정보를 불러오는 부분에서 사용했습니다! 
이 미션에서 앞에 두개의 데이터가 등록되어 있어서 임으로 START_NUMBER를 지정해서 사용했는데요. 
이 부분은 수정하겠습니다.

java-saeng commented on src/main/java/codesquad/UserController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @GetMapping()
+    public String getUserList(Model model) {
+        model.addAttribute("users", users);
```
User 단건 조회일 경우에는 UserDTO를 반환하고, 여러 User를 조회할 경우에는 User를 반환하는 것으로 보입니다~

mywnajsldkf replied:
@java-saeng 
확인감사합니다! 수정하였습니다~
다만 이번에 피드백 반영하면서 Entity->DTO, DTO->Entity로 변환하는 메서드를 어디에 두어야할지 고민되었습니다. 블로그마다 다르던데 상황에 맞게 선택하시는 편인가요?

hyukjin-lee replied:
Entity 에 DTO 변환 메서드를 두는건 확실히 지양해야할 방식이고, DTO 에서 Entity 를 변환하는 메서드를 지니는건 고려해볼만 하다고 생각이 듭니다. 따로 변환을 담당하는 converter 나 assembler 를 두어도 좋을 것 같네요.
Entity 와 DTO 형태가 크게 차이나지 않는다면 ModelMapper 나 MapStruct 같은 것도 사용해볼만 한 것 같습니다.

mywnajsldkf replied:
@hyukjin-lee 
오호 그렇군요!
형태가 차이나지 않아 ModelMapper를 사용해서 수정해보았습니다. 

이번에는 수정하면서 ModelMapper를 어느 곳에서 사용할지 생각해봤습니다. DTO를 사용하는 이유가 DB와 View 사이의 역할을 분리하고, View에서 사용하기 위한 용도이므로 service, repository에 들어가지 않고 controller에서 사용했습니다. 

다만 ModelMapper를 사용하면 맵핑하기 위해 entity에 Setter를 추가해야 합니다. entity에 setter 사용을 지양하라는 글을 많이 봤는데요. 편의를 위해 Setter를 추가해도 될지, 아니면 어떻게 생각하지는지 궁금합니다.

hyukjin-lee replied:
나중에 궁금하시다면 ModelMapper 와 MapStruct 의 장단점에 대해 보셔도 좋을 것 같습니다 ~
ModelMapper 로 인해 생기는 Setter 에 대한 걱정은 "ModelMapper Entity setter" 라고 검색해보시면 방법이 나올 것 같습니다 ~

java-saeng commented on src/main/java/codesquad/User.java:
```diff
@@ -0,0 +1,64 @@
+    private String name;
+    private String email;
+
+    public User() {}
```
기본 생성자를 만드신 이유가 있으실까요~?

mywnajsldkf replied:
@java-saeng 
안만드니 오류가 생겨서 만들었는데요... 콕 집어주신 덕분에 찾아보았습니다😅

자바에서 상속할 때 자식 클래스는 부모 클래스에서 생성된 기본 생성자에서 `super()`를 이용해 부모의 기본 생성자를 불러옵니다. 이처럼 상속하는 경우에 필요하므로 부모에서 기본 생성자를 만드는 것이 필요하다고 생각합니다!

+) 이해가 조금 부족한데 부연설명 자료 혹시 있으시다면 추가해주시면 감사하겠습니다!

hyukjin-lee replied:
User 를 상속하셨나요? + 기본 생성자를 만들지 않으면 어떤 오류가 뜨던가요?

mywnajsldkf replied:
@hyukjin-lee 
User는 상속하지 않았습니다. 
다시 확인해보니 기본 생성자를 생성한 이유는 컨트롤러의  `findUser` 메서드에서 기본 생성자를 호출하고 새로운 객체를 생성해 반환(`return new User()`)했습니다. 
이때 이미 매개변수를 가지는 생성자가 있었기 때문에 User에 기본 생성자를 생성하지 않고 사용하여 컴파일 오류가 발생했습니다.

hyukjin-lee replied:
원인을 찾으셨군요 ~ 👍 
오류상황을 그냥 넘어가거나 추측하고 끝내는 것이 아니라 원인을 파고들고 막히는 부분은 질문을 통해 문제를 해결해내는 것이 매우 중요해서 답변에서 걸리는 부분을 여쭤봤습니다 ~
여러명이 피드백을 각자의 방식대로 하다보니 혹시라도 버거운 부분이 있으면 언제든 슬랙으로 자유롭게 말씀 부탁드립니다.
정인님께서 이해도도 높고 잘 따라오는 것 같은데 노파심에 남겨봅니다 ~

java-saeng commented on src/main/java/codesquad/User.java:
```diff
@@ -0,0 +1,64 @@
+public class User {
+
+    @Id
+    private int id;
```
id가 int 형을 가질 수 있을까요~?

mywnajsldkf replied:
@java-saeng 
제가 id를 index 역할로 활용해서 int 형을 사용했습니다. 또 id는 [@Id 어노테이션](https://docs.oracle.com/javaee/6/api/javax/persistence/Id.html)은 기본형 타입과 함께 활용될 수 있기 때문에 int 형을 가질 수 있다고 생각합니다. 
현우님의 리뷰를 받고 좀더 찾아보니 id는 @Id 어노테이션과 함께 활용되는 만큼 기본키의 역할을 해야한다고 생각합니다. 
만약 기본키가 null을 정확히 처리할 수 없다면 문제가 되기 때문에 null 값을 정확하게 처리할 수 있고, 표현할 수 있는 숫자의 범위가 넓은 Long이 적절할 것 같습니다~

java-saeng replied:
이것도 같이 참고해보시면 좋을 것 같습니다 ~  (4.1.2)
https://docs.jboss.org/hibernate/orm/3.3/reference/en/html/persistent-classes.html#persistent-classes-pojo-identifier

java-saeng commented:
미션 하시느라 고생 많으셨습니다~

mywnajsldkf committed:
feat: UserDTO 생성

mywnajsldkf committed:
refactor: User 객체 Entity로 구분

mywnajsldkf committed:
feat: UserRepository 생성
- DB 접근과 관련된 동작

mywnajsldkf committed:
feat: UserService 인터페이스, UserService 구현 객체 생성
- 비즈니스 로직과 관련된 동작 수행

mywnajsldkf committed:
refactor: DTO 적용

mywnajsldkf committed:
refactor: index 자동 증가되도록 수정

mywnajsldkf committed:
refactor: indexHelper 작성

- index 시작 지점 지정

mywnajsldkf committed:
refactor: 의존성 주입 적용

mywnajsldkf committed:
chores: 불필요한 코드 삭제

mywnajsldkf committed:
refactor: modelMapper 추가하여 Dto -> entity, entity -> Dto 변환
- UserEntity Setter 추가
ㄴ ModelMapper는 해당 클래스의 기본 생성자를 이용해 객체를 생성하고 setter를 이용해 매핑함
- 불필요한 코드 제거

java-saeng commented on src/main/java/codesquad/entity/UserEntity.java:
```diff
@@ -0,0 +1,36 @@
+public class UserEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
GeneratedValue 가 무엇이고, 각 strategy 는 어떤 게 있을까요?

mywnajsldkf replied:
@java-saeng 
@GeneratedValue는 직접 기본키를 생성하는 방법입니다. 
기본키 생성 전략에는 4가지가 있습니다. 
- `@GeneratedValue(strategy = GenerationType.IDENTITY)`: id 값을 할당하지 않아도 데이터베이스가 자동으로 기본키를 생성합니다. 
- `@GeneratedValue(strategy = GenerationType.SEQUENCE)`: Sequence Object를 사용하여 자동으로 기본키를 생성합니다. (@SequenceGenerator 어노테이션 필요)
- `@GeneratedValue(strategy = GenerationType.TABLE)`: 키를 생성하는 테이블을 사용합니다. (@TableGenerator 어노테이션 필요)
- `@GeneratedValue(strategy = GenerationType.AUTO)`: 각 데이터베이스에 따라 기본키를 자동으로 생성합니다.

java-saeng replied:
현재 미션에서는 DB를 사용하지 않기 때문에 해당 어노테이션이 필요할까? 라는 생각에서 여쭤봤었습니다~

어려운 내용들 공부하시느라 고생하셨습니다

java-saeng commented on src/main/java/codesquad/service/UserService.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+public interface UserService {
```
Service와 Repository 를 interface로 만드신 이유가 궁금합니다~

mywnajsldkf replied:
@java-saeng 
다형성을 만족시키기 위해 역할과 구현을 구분하여 interface를 생성했습니다.
Service 인터페이스는 비즈니스 로직이 수행할 역할을 담고 있고, Repository 인터페이스는 데이터베이스가 수행할 역할을 담습니다. 실제 구현은 인터페이스를 구현하는 구현 객체가 담당합니다. 
이렇게 역할과 구현을 구분하여 비즈니스 로직이 변경되거나, 데이터베이스가 변경되어도 유연하게 교체할 수 있습니다. 

그런데 다형성을 만족시킬 때 상속은 적절하지 않나요?
